src/main/java/com/mathworks/ci/MatlabStepExecution.java

```diff
@@ -0,0 +1,101 @@
+package com.mathworks.ci;
+
+/**
+ * Copyright 2019-2020 The MathWorks, Inc.
+ *  
+ */
+
+import java.io.IOException;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+import hudson.EnvVars;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.Launcher.ProcStarter;
+import hudson.model.Result;
+import hudson.model.TaskListener;
+
+public class MatlabStepExecution extends StepExecution implements MatlabBuild {
+    private static final long serialVersionUID = 1L;
+    private String command;
+    private EnvVars env;
+    private boolean copyScratchFile;
+
+
+    public MatlabStepExecution(StepContext context, String command, boolean copyScratchFile) {
+        super(context);
+        this.command = command;
+        this.copyScratchFile = copyScratchFile;
+    }
+
+    private String getCommand() {
+        return this.env == null ? getMatlabCommand() : this.env.expand(getMatlabCommand());
+    }
+
+    private String getMatlabCommand() {
+        return this.command;
+    }
+
+    private void setEnv(EnvVars env) {
+        this.env = env;
+    }
+
+    private EnvVars getEnv() {
+        return this.env;
+    }
+
+    @Override
+    public boolean start() throws Exception {
+        Launcher launcher = getContext().get(Launcher.class);
+        FilePath workspace = getContext().get(FilePath.class);
+        TaskListener listener = getContext().get(TaskListener.class);
+        EnvVars env = getContext().get(EnvVars.class);
+        setEnv(env);
+
+        int res = execMatlabCommand(workspace, launcher, listener, getEnv());
+        if (res == 0) {
+            getContext().setResult(Result.SUCCESS);
+        } else {
+            getContext().setResult(Result.FAILURE);
+        }
+        getContext().onSuccess(true);
+        
+        //return false represents the asynchronous run. 
+        return false;
+    }
+
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        getContext().onFailure(cause);
+    }
+
+    private synchronized int execMatlabCommand(FilePath workspace, Launcher launcher,
+            TaskListener listener, EnvVars envVars) throws IOException, InterruptedException {
+        final String uniqueTmpFldrName = getUniqueNameForRunnerFile();
+        ProcStarter matlabLauncher;
+        try {
+            matlabLauncher = getProcessToRunMatlabCommand(workspace, launcher, listener, envVars,
+                    getCommand(), uniqueTmpFldrName);
+            
+            // Copy MATLAB scratch file into the workspace if required.
+            if(this.copyScratchFile) {
+                FilePath targetWorkspace = new FilePath(launcher.getChannel(), workspace.getRemote());
+                copyFileInWorkspace(MatlabBuilderConstants.MATLAB_TESTS_RUNNER_RESOURCE,
+                        MatlabBuilderConstants.MATLAB_TESTS_RUNNER_TARGET_FILE, targetWorkspace);
+            }
```

Comment:
This is a bit of a code smell. It seems like `MatlabStepExecution` is meant to be a general step execution class used by `RunMatlabCommandStep` and `RunMatlabTestsStep`? In that case, it should have no knowledge of a test scratch file.

I would consider removing the `copyScratchFile` stuff from this class and moving it to `RunMatlabTestsStep` if that's what needs it. Or the concept of a scratch file should be generalized beyond a test runner resource.

Reply:
Correct the reason was to avoid redundant code. However the copy was kept here to avoid the Early copy operation. I have implemented the changes as you suggested.